clang/include/clang/AST/Expr.h

```diff
@@ -4810,6 +4810,129 @@ class SourceLocExpr final : public Expr {
   friend class ASTStmtReader;
 };
 
+/// Represents a #embed "expression".
+class PPEmbedExpr final : public Expr {
+  SourceLocation BuiltinLoc, RParenLoc;
+  DeclContext *ParentContext;
+  StringLiteral *Filename = nullptr;
+  StringLiteral *BinaryData = nullptr;
+  IntegerLiteral *FakeChildNode = nullptr;
+  const ASTContext *Ctx = nullptr;
+
+public:
+  enum Action {
+    NotFound,
+    FoundOne,
+    Expanded,
+  };
+
+  PPEmbedExpr(const ASTContext &Ctx, StringLiteral *Filename,
+              StringLiteral *BinaryData, SourceLocation BLoc,
+              SourceLocation RParenLoc, DeclContext *Context);
+
+  /// Build an empty call expression.
+  explicit PPEmbedExpr(EmptyShell Empty) : Expr(SourceLocExprClass, Empty) {}
+
+  /// If the PPEmbedExpr has been resolved return the subexpression
+  /// representing the resolved value. Otherwise return null.
+  const DeclContext *getParentContext() const { return ParentContext; }
+  DeclContext *getParentContext() { return ParentContext; }
+
+  SourceLocation getLocation() const { return BuiltinLoc; }
+  SourceLocation getBeginLoc() const { return BuiltinLoc; }
+  SourceLocation getEndLoc() const { return RParenLoc; }
+
+  StringLiteral *getFilenameStringLiteral() const { return Filename; }
+  StringLiteral *getDataStringLiteral() const { return BinaryData; }
+
+  size_t getDataElementCount(ASTContext &Context) const;
+
+private:
+  template <bool Const>
+  class ChildElementIter
+      : public llvm::iterator_facade_base<
+            // FIXME: it seems reasonable to make this a random access iterator
+            // instead, but all current access patterns are a linear walk over
+            // the contents, so it's being left for follow-up work if needed.
+            ChildElementIter<Const>, std::input_iterator_tag,
+            std::conditional_t<Const, const IntegerLiteral *,
+                               IntegerLiteral *>> {
+    friend class PPEmbedExpr;
+
+    const ASTContext *Ctx = nullptr;
+    IntegerLiteral *FakeNode = nullptr;
+    StringRef DataRef;
+    LLVM_PREFERRED_TYPE(bool) unsigned IsSigned : 1;
```

Comment:
Not a huge advantage, no. We could also use a `PointerIntUnion` for `FakeNode` and `IsSigned` as a way to save that bit. It mostly just offends my sensibilities to take up that much memory for a single bit of information. :-D

Do you want me to go with `bool` and `unsigned`, `PointerIntUnion` and `unsigned`, or leave as-is?

Reply:
My point is the field after (curOffset) this could be 63 bits and we'd lose nothing, since this file (on 64 bit builds) has 32 'padding' bits as written.

Reply:
Oh thank you for clarifying, I hadn't picked up on that! I'll switch to 63 bits instead of 31. However... I don't think I want to test that scenario because that would require folks clone a rather large file (or have lit generate one on the fly). You comfortable with that?

Reply:
The other option is to jsut remove bitfields entirely here, and just do a 'bool' and a '32 bit'. I don't mind being unable to test the situation, since it is basically impossible to run into on any modern OS (a file bigger than 63 bits in length).

If we wanted to be EXTRA clever we could do a `#if` around the 31/63 based on whether this is a 32 or 64 bit machine (or based on `sizeof(size_t)`?), then only enable the 'too big' test on 32 bit machines?

Reply:
I think I'm happy enough with using a 63-bit int here; it's certainly possible to have a file whose size is greater than 31 bits. It may waste a bit of extra space on 32-bit systems, but I don't imagine `#embed` showing up so often in the AST for it to matter.

Reply:
This ended up being obviated somewhat -- I went with `unsigned long long` in this iteration and the bit-field and other pointers are gone.